### ISSUE TITLE
When there is only one open repository but lots of closed ones, WrongNumberOfRepositories exception is thrown

### ISSUE THREAD

**Issue by strelok1** [Codearte/gradle-nexus-staging-plugin#36]
Or am I doing something wrong? 

io.codearte.gradle.nexus.logic.RepositoryFetcher#getOpenRepositoryIdForStagingProfileId method is called with ("open", stagingProfileId) and following the call stack it eventually reaches the 

io.codearte.gradle.nexus.logic.RepositoryFetcher#verifyThatOneRepositoryAndReturnIt method which takes the response map and the `repositoryState` parameter. This parameter is never used. I have hundreds of "closed" repositories in a staging profile, but only one "open" one. This method of course throws a WrongNumberOfRepositories exception as it checks the response size only.

Shouldn't it be:

```
    private Map verifyThatOneRepositoryAndReturnIt(Map responseAsMap, String repositoryState) {
        int numberOfRespositories = responseAsMap.data.count { it.type == repositoryState }
        if (numberOfRespositories != 1) {
            throw new WrongNumberOfRepositories(numberOfRespositories, repositoryState)
        }
        Map repository = responseAsMap.data.find { it.type == repositoryState } as Map
        return repository
    }

```

People using this plugin, do you ever only have ONE repository in your staging profile and it's always open?

**Comment by szpak** [comment]
It sounds sensible to filter repositories by state in that method. Would you like to provide a PR with a fix (and an automatic test for that)?

Btw, are those "hundreds" of closed repositories generated by the plugin or you have some other releasing activity causing that? In the first case #37 would be beneficial.

**Comment by strelok1** [comment]
@szpak it's other deployment activity, not the plugin. I will provide a PR. I am having a hard time getting the tests to pass though. Just running `./gradlew test` after cloning the repo causes 14 failed tests. Things like: 

```
io.codearte.gradle.nexus.functional.MockedFunctionalSpec > should call close and promote in closeAndPromoteRepository task FAILED
    org.gradle.api.GradleException at MockedFunctionalSpec.groovy:174
        Caused by: org.gradle.internal.exceptions.LocationAwareException
            Caused by: org.gradle.api.GradleScriptException
                Caused by: org.gradle.api.plugins.UnknownPluginException
```

Am I doing something wrong?

**Comment by szpak** [comment]
There was one small issue with functional tests - #31, but it was fixed. Travis builds current master without [any problem](https://travis-ci.org/Codearte/gradle-nexus-staging-plugin/builds/194894653). 

1. Could you re-run tests with `--info --stacktrace` to see if there is anything more meaningful?
2. What OS do you use?

**Comment by strelok1** [comment]
@szpak Sample stack trace;

```
Running test from C:\code\gradle-nexus-staging-plugin\build\test\io.codearte.gradle.nexus.functional.MockedFunctionalSpec\should-not-do-request-for-staging-profile-when-provided-in-configuration-on-testedTaskName-task_8

org.gradle.api.GradleException: Build aborted because of an internal error.

	at nebula.test.functional.internal.DefaultExecutionResult.rethrowFailure(DefaultExecutionResult.groovy:97)
	at nebula.test.IntegrationSpec.runTasksSuccessfully(IntegrationSpec.groovy:265)
	at io.codearte.gradle.nexus.functional.MockedFunctionalSpec.should not do request for staging profile when provided in configuration on #testedTaskName task(MockedFunctionalSpec.groovy:36)
Caused by: org.gradle.internal.exceptions.LocationAwareException: Build file 'C:\code\gradle-nexus-staging-plugin\build\test\io.codearte.gradle.nexus.functional.MockedFunctionalSpec\should-not-do-request-for-staging-profile-when-provided-in-configuration-on-testedTaskName-task_8\build.gradle' line: 4
A problem occurred evaluating root project 'should-not-do-request-for-staging-profile-when-provided-in-configuration-on-testedTaskName-task'.
	at org.gradle.initialization.DefaultExceptionAnalyser.transform(DefaultExceptionAnalyser.java:74)
	at org.gradle.initialization.MultipleBuildFailuresExceptionAnalyser.transform(MultipleBuildFailuresExceptionAnalyser.java:47)
	at org.gradle.initialization.StackTraceSanitizingExceptionAnalyser.transform(StackTraceSanitizingExceptionAnalyser.java:30)
	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:101)
	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:93)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:62)
	at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:93)
	at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:82)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:94)
	at org.gradle.tooling.internal.provider.runner.BuildModelActionRunner.run(BuildModelActionRunner.java:46)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.tooling.internal.provider.runner.SubscribableBuildActionRunner.run(SubscribableBuildActionRunner.java:58)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:43)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:28)
	at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:81)
	at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:46)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:52)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
	at org.gradle.util.Swapper.swap(Swapper.java:38)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.health.DaemonHealthTracker.execute(DaemonHealthTracker.java:47)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:60)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:72)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.health.HintGCAfterBuild.execute(HintGCAfterBuild.java:41)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:246)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
Caused by: org.gradle.api.GradleScriptException: A problem occurred evaluating root project 'should-not-do-request-for-staging-profile-when-provided-in-configuration-on-testedTaskName-task'.
	at org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:93)
	at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl$2.run(DefaultScriptPluginFactory.java:157)
	at org.gradle.configuration.ProjectScriptTarget.addConfiguration(ProjectScriptTarget.java:72)
	at org.gradle.configuration.DefaultScriptPluginFactory$ScriptPluginImpl.apply(DefaultScriptPluginFactory.java:162)
	at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:38)
	at org.gradle.configuration.project.BuildScriptProcessor.execute(BuildScriptProcessor.java:25)
	at org.gradle.configuration.project.ConfigureActionsProjectEvaluator.evaluate(ConfigureActionsProjectEvaluator.java:34)
	at org.gradle.configuration.project.LifecycleProjectEvaluator.evaluate(LifecycleProjectEvaluator.java:55)
	at org.gradle.api.internal.project.AbstractProject.evaluate(AbstractProject.java:529)
	at org.gradle.api.internal.project.AbstractProject.evaluate(AbstractProject.java:90)
	at org.gradle.execution.TaskPathProjectEvaluator.configureHierarchy(TaskPathProjectEvaluator.java:42)
	at org.gradle.configuration.DefaultBuildConfigurer.configure(DefaultBuildConfigurer.java:35)
	at org.gradle.initialization.DefaultGradleLauncher$2.run(DefaultGradleLauncher.java:125)
	at org.gradle.internal.Factories$1.create(Factories.java:22)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:52)
	at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:122)
	at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:32)
	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:99)
	... 42 more
**Caused by: org.gradle.api.plugins.UnknownPluginException: Plugin with id 'io.codearte.nexus-staging' not found.**
	at org.gradle.api.internal.plugins.DefaultPluginManager.apply(DefaultPluginManager.java:110)
	at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.applyType(DefaultObjectConfigurationAction.java:113)
	at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.access$200(DefaultObjectConfigurationAction.java:36)
	at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction$3.run(DefaultObjectConfigurationAction.java:80)
	at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.execute(DefaultObjectConfigurationAction.java:136)
	at org.gradle.api.internal.project.AbstractPluginAware.apply(AbstractPluginAware.java:46)
	at org.gradle.api.internal.project.ProjectScript.apply(ProjectScript.groovy:35)
	at build_cw6cuz34b890se4thjgcstlxv.run(C:\code\gradle-nexus-staging-plugin\build\test\io.codearte.gradle.nexus.functional.MockedFunctionalSpec\should-not-do-request-for-staging-profile-when-provided-in-configuration-on-testedTaskName-task_8\build.gradle:4)
	at org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:91)
	... 60 more
```

Root cause is always the same "Plugin with id 'io.codearte.nexus-staging' not found."

Windows 7 SP1 64-bit, Oracle JDK.

**Comment by strelok1** [comment]
Doesn't matter, travis worked for me too. I will just use that, rather then running locally. :+1: Cloud CI.

**Comment by szpak** [comment]
The problem could be related to https://github.com/nebula-plugins/nebula-test/pull/77. I created a separate branch for you - [tech/windowsCompatibility](https://github.com/Codearte/gradle-nexus-staging-plugin/tree/tech/windowsCompatibility) - with new nebula-test (which is not a simple nebula bump as could be expected). You can give it a try.

**Comment by strelok1** [comment]
Yep. That fixed it for me. I had to upgrade the wrapper to 3.4 though (2.4 has a bug with version comparison and it would not take an newer version of Animal Sniffer) and include guava as a "compile" dependency because of usage of VisibleForTesting. Before it was included as part of nebula-test and obviously old gradle versions didn't really care about actual testCompile scope;)

The problem with my fix ofcourse is that promoteRepository does not work still ,ofcourse as there are hundreds of "closed" repositories in my case and it can't find JUST ONE. I think I can fix it for the case when close and promote are called together (by saving the repository Id in the close task like you do with the staging profile Id). I think this will fix the issue in my case, as I'll be ok with closing and promoting right away I guess while still maintaining compatibility for people that are lucky to only have just one closed repository. What do you think of this fix?

**Comment by strelok1** [comment]
#38

**Comment by strelok1** [comment]
@szpak take a look at https://github.com/Codearte/gradle-nexus-staging-plugin/pull/38

It didn't mention this issue by accident :)

**Comment by szpak** [comment]
Available in 0.6.0. Thanks.
